storage/innobase/log/log0recv.cc

```diff
@@ -2946,18 +2955,6 @@ recv_sys_t::parse_mtr_result recv_sys_t::parse(source &l, bool if_exists)
           }
         }
       }
-      else if ((b & 0x70) <= INIT_PAGE)
-      {
-        mlog_init.add(id, start_lsn);
-        if (pages_it == pages.end() || pages_it->first != id)
-        {
-          pages_it= pages.find(id);
-          if (pages_it == pages.end())
-            continue;
-        }
-        map::iterator r= pages_it++;
-        erase(r);
-      }
     }
```

Comment:
This matched `INIT_PAGE` and `FREE_PAGE` in the template expansion `bool store=false`. The code was redundant, because the same records will be guaranteed to be parsed by a subsequent invocation of `recv_scan_log()` that will use `store=true`.

In `mariadb-backup --backup` (`storing == BACKUP` with this change) this code is completely unnecessary. In recovery during `storing == NO` we do not need this either. This would be during either the `skip_the_rest:` loop or the initial parsing of the `FILE_CHECKPOINT` record. Yes, invoking `erase(r)` could free up some space from `recv_sys.pages` during that batch, but it would not have any benefit; nothing is going to add anything to `recv_sys.pages`. This logic will be executed as part of the preceding `if (storing == YES)` template expansion branch, in the invocation of `recv_sys_t::add()`:
```c++
  switch (*l & 0x70) {
  case FREE_PAGE: case INIT_PAGE:
    recs.will_not_read();
    mlog_init.add(it->first, start_lsn);
```
Note: the `recs.will_not_read()` is a ‘lazy’ equivalent of the `erase(r)` in the above code snippet. The idea is that if there are subsequent log records for the same page, we will preserve the `std::map` structure (not delete or add keys), only modify the ‘value’ in place.

Reply:
Sorry, I realized that the `erase()` is actually needed, both for `INIT_PAGE` and `FREE_PAGE`. I debugged a failure of the test `innodb.undo_truncate_recover`, where we were reading an all-NUL page for which there had been some log records before a `FREE_PAGE` record had been written.

Reply:
It turns out that also the `mlog.init.add()` will be needed also for `storing == NO`. Otherwise, the test `innodb.recovery_memory` and some other tests that may invoke multi-batch recovery (such as `innodb_gis.types`) may occasionally fail.

Reply:
b249a059da2ef03eda35872a763e5453dbc39339 retains all this logic for `storing == NO`. With that, the test `innodb.recovery_memory` never reports corruption for me.